### DISCUSSION TITLE
Flip the display & Pixelit/Awtrix Font

### DISCUSSION THREAD

**Discussion by trip5** [lubeda/EsphoMaTrix#78] [Ideas]
How hard would it be be to implement a 180 flip of the display? The Ulanzi TC001 case has an angled top which is almost perfect (needs a thick USB cable to support it better) so that the display can sit lower than eye-level under my TV (wife-approval factor for the old Awtrix in the living room).

Or is the functionality already there in the pixel-mapping in some way?

**Comment by lubeda** [comment]
Hi,
just check if the rotation parameter under the display key does the job. Pixelmapper would work also but needs more testing.

**Reply by trip5** [reply]
Thanks so much.  Lol, it's right there in my YAML already (my base code being from https://blakadder.com/esphome-pixel-clock/).  Thanks so much!  I guess I hadn't scrolled down far enough!

**Comment by lubeda** [comment]
No problem, please consider giving me a github star!

**Comment by trip5** [comment]
Done.  I've even adapted my YAML to "backport" to my old Awtrix already. Love the ability to use ESPHome in a clock with your code!

The only think I miss about the Awtrix is Blueforcers' modified TomThumb font.  The original appears to be GPL...  I'll come back again if I can figure out how to get it into a TTF properly.

**Comment by lubeda** [comment]
That would be very helpfull, i din't found a way to convert it to ttf

**Comment by trip5** [comment]
I'm actually making some progress on this.  I found the original bdf file that Blueforcer & all the others appear to have used... running it through Fontforge and editing it now... Since I no longer have the Awtrix firmware to compare (whoops), I'm using screenshots on Google but mostly my own judgment to convert the original 6x4 fixed-width font to a more variable width (like the others did).

I was debating whether or not to beg Blueforcer or Pixelit to share their original BDF file but lol the chain of attribution is long enough for them already.  This way should be... a bit less?  Lol.  I'll get back to you when I'm satisfied.  The accented and extended characters are probably going to be troublesome.

**Reply by trip5** [reply]
Alright, alright.  After some screwing around in FontForge, I think I have some options similar as Pixelit or Awtrix.

I'm going to make a series of fonts especially for use with this project.  ^^  It'd be pretty cool if you added a link.  I think BDF fonts are much more predictable on the matrix display.

https://github.com/darkpoet78/MatrixClockFonts

**Comment by lubeda** [comment]
Hi I added your link yesterday, but I did not try your fonts yet. Do you have an euro € sign in your font? And what's about Umlauts ÖÄÜöäü for German people?

**Reply by trip5** [reply]
The first 3 fonts do not.  At 5 pixels tall, any attempt at diacritics/accents is ugly as sin and bloats the size of the font quickly.  Let me show you my workshop:

![image](https://user-images.githubusercontent.com/19512466/230439498-aef14d31-0ac4-4b12-8171-f790891c530e.png)

With a font like this would it better to at least map all the letters with accents to un-accented letters (so at least there's something on the matrix) or go with these ugly things?  Look at the E's in particular... ugh.

I'm working on a 7-pixel font now.  That should be a bit easier to work in the accents.

**Comment by lubeda** [comment]
Very nice, i tried it by my self [gist](https://gist.github.com/lubeda/7629a1f6ef9a08baa852eb1b68d08e31) but the euro sign is still a problem because of the latin-1 encoding. At least the mapping to unaccented is a way to go but the Euro sign is a "must".

**Comment by trip5** [comment]
Played around quite a bit with a Fontforge tonight in a 7 pixel height... It looks a lot better at that height, except some of those Spanish and French things.  Lol.  It doesn't seem to be a problem to encode various currency symbols, but I'll get back to you after testing.

I noticed the Bitcoin symbol just below the Euro symbol, too... I'll see if I can add a few of those to the larger fonts then go back for the smaller fonts.

**Comment by trip5** [comment]
Figured I'd update you.  Got a bit annoyed completing the entire European set of accents.  Gosh, there are marks I didn't even know existed!

But it's done (for now).  ^^

OMG Spoke too soon.  Something is wrong with my encoding, I think.  I can't get my device to show this little test "45€ hôtel" through Home Assistant - argh.  I'll see what's up over the weekend if I can.

**Comment by lubeda** [comment]
Very nice, i am working on the 2023.5.0. There are a lot of internal changes and cleanup. But also i will support two fonts, a default and a small font. You will be able to select the font per screen, so you can use the 6 pixel fonts with the clock and the 8 pixel fonts for the rest.

But this redesign will include a lot of breaking changes, but i think they are worth the effort.

**Reply by trip5** [reply]
I'm looking forward to that!

By the way, it looks like when using a BDF font, ESPHome's python dependency Pillow does not recognize a unicode font correctly.  I tried multiple solutions, and it doesn't help that ESPHome is using an older version of Pillow as a system python wheel... so I couldn't even upgrade the wheel inside my docker container...

Tried various headers in my BDF file to get Pillow to recognize the Unicode.  It continued to assume it was Latin-1.  Tried forcing Pillow into UTF-8 mode.  Ended up with a double-byte error (2 characters showed up on my clock instead of 1).

So, it looks like with BDF files, we're stuck with the Latin-1 limitation.  It should be enough...?  In lieu of an actual solution to the problem, I turned some "useless" characters into useable currency characters for your Euro and my Won.  I'm not happy with the end result, but it'll do.
